src/RenderAPI/EspPipeline.cc

```diff
@@ -0,0 +1,293 @@
+#include "EspPipeline.hh"
+#include "Scene/Components/Components.hh"
+
+// std
+#include <fstream>
+#include <stdexcept>
+#include <iostream>
+#include <cassert>
+
+#ifndef ENGINE_DIR
+#define ENGINE_DIR "../"
+#endif
+
+namespace esp
+{
+	EspPipeline::Builder& EspPipeline::Builder::set_vert_shader_path(const std::string& path)
+	{
+		m_vert_shader_path = path;
+		return *this;
+	}
+
+	EspPipeline::Builder& EspPipeline::Builder::set_frag_shader_path(const std::string& path)
+	{
+		m_frag_shader_path = path;
+		return *this;
+	}
+
+	std::unique_ptr<EspPipelineLayout> EspPipeline::Builder::build_pipeline_layout(
+		PipelineConfigInfo& pipeline_config)
+	{
+		VkPipelineLayoutCreateInfo pipelineLayoutCreateInfo{};
+		pipelineLayoutCreateInfo.sType = VK_STRUCTURE_TYPE_PIPELINE_LAYOUT_CREATE_INFO;
+		pipelineLayoutCreateInfo.setLayoutCount = 0;
+		pipelineLayoutCreateInfo.pSetLayouts = nullptr;
+		pipelineLayoutCreateInfo.pushConstantRangeCount = 0;
+		pipelineLayoutCreateInfo.pPushConstantRanges = nullptr;
+
+		VkPipelineLayout pipeline_layout{};
+		if (vkCreatePipelineLayout(m_device.get_device(), &pipelineLayoutCreateInfo, nullptr, &pipeline_layout) != VK_SUCCESS)
+		{
+			ESP_CORE_ERROR("Failed to create pipeline layout");
+			throw std::runtime_error("Failed to create pipeline layout");
+		}
+
+		pipeline_config.m_pipeline_layout = pipeline_layout;
+
+		return std::make_unique<EspPipelineLayout>(pipeline_layout);
+	}
+
+	std::unique_ptr<EspPipeline> EspPipeline::Builder::build_pipeline(
+		esp::PipelineConfigInfo& pipeline_config,
+		VkRenderPass render_pass)
+	{
+		assert(pipeline_config.m_pipeline_layout != nullptr
+			&& "Can't create pipeline before pipeline layout");
+
+		pipeline_config.m_render_pass = render_pass;
+
+		return std::make_unique<EspPipeline>(
+			m_device,
+			m_vert_shader_path,
+			m_frag_shader_path,
+			pipeline_config);
+	}
+
+	EspPipeline::EspPipeline(
+		EspDevice& device,
+		const std::string& shader_vert_path,
+		const std::string& shader_frag_path,
+		const PipelineConfigInfo& config_info
+	) : m_device{ device }
+	{
+		create_graphics_pipeline(shader_vert_path, shader_frag_path, config_info);
+	}
+
+	EspPipeline::~EspPipeline()
+	{
+		vkDestroyShaderModule(m_device.get_device(), m_vert_shader_module, nullptr);
+		vkDestroyShaderModule(m_device.get_device(), m_frag_shader_module, nullptr);
+		vkDestroyPipeline(m_device.get_device(), m_graphics_pipeline, nullptr);
+	}
+
+	std::vector<char> EspPipeline::read_file(const std::string& file_path)
+	{
+		std::string engine_path = ENGINE_DIR + file_path;
```

Comment:
Yeah, thanks

Reply:
But I think it would be much more convenient to let user define an env variable for this path. Perhaps we could then pass it to core. You can propose a solution for that if you want:)